File: 0275/binary_search.cpp
#include <cstdlib>
#include <vector>

#include "gtest/gtest.h"

using namespace std;

class Solution {
 public:
  int hIndex(vector<int>& citations) {
    int left = 0, right = citations.size()-1, len = citations.size();
    while (left <= right) {
      int mid = (right - left) / 2 + left;
      // `num` is the number of values greater than or equal to `citations[mid]`
      int num = len - mid;
      if (num > citations[mid]) {
        left = mid + 1;
      } else if (num < citations[mid]) {
        right = mid - 1;
      } else {
        return citations[mid];
      }
    }
    // TODO: why?
    return len - (right+1);
  }
};

int main() {
  Solution sol;
  vector<int> citations;
  int expected;

  citations = {0,1,3,5,6};
  expected = 3;
  EXPECT_EQ(expected, sol.hIndex(citations));

  citations = {0,1,2,2,2,2,2};
  expected = 2;
  EXPECT_EQ(expected, sol.hIndex(citations));

  citations = {0,1,2,4,5,5,5};
  expected = 4;
  EXPECT_EQ(expected, sol.hIndex(citations));

  citations = {2,5,5,5,5,5};
  expected = 5;
  EXPECT_EQ(expected, sol.hIndex(citations));

  citations = {0};
  expected = 0;
  EXPECT_EQ(expected, sol.hIndex(citations));

  citations = {1};
  expected = 1;
  EXPECT_EQ(expected, sol.hIndex(citations));

  citations = {2};
  expected = 1;
  EXPECT_EQ(expected, sol.hIndex(citations));

  exit(EXIT_SUCCESS);
}

File: 1552/binary_search.py
from typing import List


class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:
        # If the number of balls is greater than or equal to `m`,
        # a distance between two balls can be bigger.
        def isOk(d):
            prev_idx = 0
            num_balls = 1
            for idx in range(1, len(position)):
                if position[idx] - position[prev_idx] >= d:
                    num_balls += 1
                    prev_idx = idx
            return num_balls >= m

        position.sort()
        # `ok` and `ng` are a minimum distance between two balls.
        ok, ng = 0, position[-1] - position[0] + 1
        while (ng - ok) > 1:
            mid = int((ng - ok) / 2) + ok
            if isOk(mid):
                ok = mid
            else:
                ng = mid
        return ok


if __name__ == '__main__':
    sol = Solution()

    position, m = [1, 2, 3, 4, 7], 3
    expected = 3
    assert sol.maxDistance(position, m) == expected

    position, m = [5, 4, 3, 2, 1, 1000000000], 2
    expected = 999999999
    assert sol.maxDistance(position, m) == expected
